src/stdx/zipfian.zig

```diff
@@ -0,0 +1,490 @@
+//! deck randomly, preserving the relative probability of existing keys.
+//!
+//! The non-shuffled generator should pass a 2-sample Kolmogorov–Smirnov test;
+//! the shuffled generator does not because the tail distribution is fudged.
+
+const std = @import("std");
+const assert = @import("std").debug.assert;
+const Random = @import("std").Random;
+const math = @import("std").math;
+const BoundedArray = @import("./bounded_array.zig").BoundedArray;
+
+/// The default "skew" of the distribution.
+const theta_default = 0.99; // per YCSB
+
+/// Generates Zipfian-distributed numbers from 0 to a specified maximum.
+///
+/// The internal variables here are the same is in the paper;
+/// the external intended to be more understandable to the user.
+pub const ZipfianGenerator = struct {
+    const Self = @This();
+
+    theta: f64,
+
+    /// The number of items in the set.
+    n: u64,
+    /// The Riemann zeta function calculated up to `n`,
+    /// aka the "generalized harmonic number" of order `theta` for `n`.
+    /// This is a pre-calculated factor in the probability of any particular item
+    /// being selected.
+    /// It is expensive to calculate for large but useful values of `n`,
+    /// but can be calculated incrementally as `n` grows.
+    zetan: f64,
+
+    /// Create a generator from `[0, items)` with `theta` equal to 0.99.
+    pub fn init(items: u64) ZipfianGenerator {
+        return ZipfianGenerator.init_theta(items, theta_default);
+    }
+
+    /// Create a generator from `[0, items)` with given `theta`.
+    ///
+    /// `theta` is the "skew" and is usually specified to be greater than 0 and less than 1,
+    /// with YCSB using 0.99, though values greater than 1 also seem to generate reasonable
+    /// distributions.
+    pub fn init_theta(items: u64, theta: f64) ZipfianGenerator {
+        assert(theta > 0.0);
+        assert(theta != 1.0); // 1.0 does not behave reasonably.
+        return ZipfianGenerator{
+            .theta = theta,
+            .n = items,
+            .zetan = zeta(items, theta),
+        };
+    }
+
+    pub fn next(self: *const Self, rng: Random) u64 {
+        assert(self.n > 0);
+
+        // Math voodoo, copied from the paper,
+        // which doesn't explain it, but claims it is from Knuth volume 3.
+
+        const nf: f64 = @floatFromInt(self.n);
+
+        // NB: These depend only on zetan and could be cached for a minor speedup.
+        const alpha = 1.0 / (1.0 - self.theta);
+        const eta = (1.0 - math.pow(f64, 2.0 / nf, 1.0 - self.theta)) /
+            (1.0 - zeta(2.0, self.theta) / self.zetan);
+
+        const u = rng.float(f64);
+        const uz = u * self.zetan;
+
+        if (uz < 1.0) {
+            return 0;
+        }
+
+        if (uz < 1.0 + math.pow(f64, 0.5, self.theta)) {
+            return 1;
+        }
+
+        return @as(u64, @intFromFloat(nf * math.pow(f64, eta * u - eta + 1.0, alpha)));
```

Comment:
Make the precedence explicit:

```suggestion
        return @as(u64, @intFromFloat(nf * math.pow(f64, (eta * u) - eta + 1.0, alpha)));
```